### DISCUSSION TITLE
Some building walls are invisible when extrusion zoom-min < 15

### DISCUSSION THREAD

**Discussion by wolfgang-ch** [mapsforge/vtm#927] [General]
When the zoom-min value is below 15 then the building walls on some buildings are invisible

![image](https://user-images.githubusercontent.com/1283445/167584139-f4cf6fc4-a77d-43a3-ab31-cab759cc541a.png)

**zoom-min=14**

![image](https://user-images.githubusercontent.com/1283445/167584376-f8049587-cc17-4665-9b12-876ee859e671.png)

**zoom-min=15**

![image](https://user-images.githubusercontent.com/1283445/167584468-af71f1ea-d337-48dd-a701-f0f968374ceb.png)

**Comment by devemux86** [comment]
What platform?
What graphics card and driver?

What map source?

Please test with VTM samples where everyone can try to reproduce the report.

You can try without shadows (default)
and see if it works better with your graphics card.

**Reply by wolfgang-ch** [reply]
```
[LWJGL Application] INFO org.oscim.gdx.GdxMap - Type: OpenGL
Version: 4:6:0
Vendor: NVIDIA Corporation
Renderer: Quadro T1000/PCIe/SSE2

```
**Platform** : Windows 10
**Map source** : Mapilion
**Theme** : OpenMapTiles (with modified zoom-min value in /vtm-themes/resources/assets/vtm/openmaptiles.xml)

>You can try without shadows (default)

Without shadow it do not look better

It also depends on the tile, some tiles are OK some not

![image](https://user-images.githubusercontent.com/1283445/167591119-5de1aced-3dde-44e1-96aa-faaaaafce26a.png)

>Please test with VTM samples where everyone can try to reproduce the report.

I need to prepare the samples

**Comment by devemux86** [comment]
[MapilionMvtTest](https://github.com/mapsforge/vtm/blob/master/vtm-playground/src/org/oscim/test/MapilionMvtTest.java) on Linux seems to work correctly.

It is recommended to use the default building min zoom level 17.
We do not know at what zoom map providers have building data.

And also try Mapsforge offline maps.

```java
mMap.layers().add(new BuildingLayer(mMap, l, 15, mMap.viewport().getMaxZoomLevel(), false, false));
```

![VTM](https://user-images.githubusercontent.com/3484020/167597278-2dd8e32a-b2fb-4708-adc8-8d45bc54b4d1.png)

**Reply by wolfgang-ch** [reply]
New York: When using the zoom level ... 17, 16 and 15 then buildings are OK but with 14 some tiles are not OK

**Zoomlevel 15**

![image](https://user-images.githubusercontent.com/1283445/167776191-9d1e9309-da34-4a62-8ea0-2b77fa3861d0.png)


**Zoomlevel 14**

![image](https://user-images.githubusercontent.com/1283445/167776253-8ab761c8-3eb3-4b7a-976e-cd9dfe8fbf3c.png)


Bremen: Zoomlevels 14 and 15 are OK

**Zoomlevel 15**

![image](https://user-images.githubusercontent.com/1283445/167777645-ff5ade3c-03cc-4790-a7e7-2e50878bf99f.png)


**Zoomlevel 14**

![image](https://user-images.githubusercontent.com/1283445/167777657-15640597-a620-4db3-acd0-e5549cfce686.png)


Paris: Looks more ugli with zoomlevel 14

**Zoomlevel 15**

![image](https://user-images.githubusercontent.com/1283445/167779378-44b9133b-80df-450d-8da4-40c62a8361ab.png)


**Zoomlevel 14**

![image](https://user-images.githubusercontent.com/1283445/167779412-d694a65a-c7cb-443b-8516-199bd74572cc.png)

**Comment by wolfgang-ch** [comment]
Mapsforge has the same issues on zoomlevel 14

![image](https://user-images.githubusercontent.com/1283445/167783856-1ac15aea-5c69-448c-9a34-85d425aa1113.png)

**Comment by devemux86** [comment]
It is recommended to use the default building min zoom level 17.

[tag-mapping](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map-writer/src/main/config/tag-mapping.xml) has `zoom-appear="17"` for the building extrusions.

**Comment by wolfgang-ch** [comment]
I would like to allow the user to set the min zoom level for buildings to the zoom levels 16 + 15 which are working very smoothly without any issues.

To do this, the `zoom` field in 
https://github.com/mapsforge/vtm/blob/4fb99aaf2811a3c1f94fb918d4c6492c9c8ad5b3/vtm/src/org/oscim/theme/rule/Rule.java#L55
needs to be set, but currently it is `final`.

I could create a test code successfully with a `RuleVisitor` to set this field in my dev environment. 

I would like to create a PR which can set this `zoom` field and it is removing the `final` keyword. 

Would you accept such a PR?
When not, which solution would you suggest?

**Comment by devemux86** [comment]
See [above](https://github.com/mapsforge/vtm/discussions/927#discussioncomment-2720989), the theme and `BuildingLayer` should work.

**Reply by wolfgang-ch** [reply]
Now I'm going the hardcore way...

![image](https://user-images.githubusercontent.com/1283445/168029450-02d35f94-34c7-44af-b813-f96c2de5af42.png)

...and it works perfectly

![image](https://user-images.githubusercontent.com/1283445/168029595-08d55431-db8e-4a41-9c0f-e7f344cb3d20.png)
